File: pharmacy-go/internal/security/authz/service.go
package authz

import (
	"context"
	"errors"

	"pharmacy-go/internal/security"
)

type Service struct {
	enforce bool
	policy  map[string]map[string]struct{}
}

func New(enforce bool) *Service {
	return &Service{
		enforce: enforce,
		policy: map[string]map[string]struct{}{
			security.ActionReadPrescription:    set("admin", "pharmacist", "doctor"),
			security.ActionScanPrescription:    set("admin", "pharmacist", "scanner"),
			security.ActionScanMedicine:        set("admin", "pharmacist", "scanner"),
			security.ActionDispensePrescription: set("admin", "pharmacist"),
		},
	}
}

func (s *Service) Authorize(_ context.Context, subject security.Subject, action string) error {
	if !s.enforce {
		return nil
	}
	allowed, ok := s.policy[action]
	if !ok {
		return errors.New("action is not allowed by policy")
	}
	for _, role := range subject.Roles {
		if _, exists := allowed[role]; exists {
			return nil
		}
	}
	return errors.New("access denied")
}

func set(roles ...string) map[string]struct{} {
	out := make(map[string]struct{}, len(roles))
	for _, role := range roles {
		out[role] = struct{}{}
	}
	return out
}

File: laboratory-python/security/authz/service.py
from ..contracts import Action, Subject


class AuthzService:
    def __init__(self, enforce: bool = False):
        self._enforce = enforce
        self._policy: dict[str, set[str]] = {
            "investigation:list_ordered": {"admin", "lab_tech", "doctor"},
            "investigation:read": {"admin", "lab_tech", "doctor"},
            "sample:register": {"admin", "lab_tech"},
            "sample:to_storage": {"admin", "lab_tech"},
            "sample:to_analysis": {"admin", "lab_tech"},
            "analyzer:create": {"admin", "tech"},
            "workstation:create": {"admin", "tech"},
            "analyzer_result:create": {"admin", "lab_tech"},
            "investigation:complete": {"admin", "lab_tech", "doctor"},
            "equipment:create": {"admin", "tech"},
            "monitoring:add_metric": {"admin", "tech"},
            "diagnostic:add": {"admin", "tech"},
        }

    def authorize(self, subject: Subject, action: Action) -> None:
        if not self._enforce:
            return
        allowed_roles = self._policy.get(action.name)
        if not allowed_roles:
            raise PermissionError("action is not allowed by policy")
        for role in subject.roles:
            if role in allowed_roles:
                return
        raise PermissionError("access denied")
